usr/sbin/p11sak/p11sak.c

```diff
@@ -1086,7 +1099,12 @@ static const struct p11sak_opt p11sak_generate_key_opts[] = {
                      "letters. For asymmetric keys set individual key "
                      "attributes for public and private key separated by a "
                      "colon (':'): 'PUB-ATTRS:PRIV-ATTRS'.", },
-    { .short_opt = 'i', .long_opt = "id", .required = false,
+    { .short_opt = 'i', .long_opt = "id",
+#if defined(_AIX)
+        .required = true,
```

Comment:
I don't really understand why this is required....

You mention this in the commit message but I don't get it still...

_p11sak requires ID for all generation operations on AIX. If the ID is not
specified, future query operations error out because the ID is not tagged at
the time of generation. This is a stop-gap measure and will be rectified in a
future update._

Reply:
This issue would cause the PKCS#11 provider that Holger developed to discard a matched object URI because the ID would be missing in the returned object. I do need to revisit this change.

Reply:
OK, sure certain applications (or providers) do need an CKA_ID attribute to be set. But thats not something that can case p11sak to always require that attribute. At the end its the PKCS#11 standard that defines if an attribute is mandatory or not. And clearly, CKA_ID is optional (i.e. it defaults to empty). 
So please revert this change, we do not want p11sak to make the CKAID mandatory.

Reply:
Oh yeah yeah, that's already done. I've removed the hunks that mandated the ID. I was just finishing with the investigation that I had to drop early the last time.

> Where exactly is that malloc with zero length ? In opencryptoki? Or in the provider ?

Not within opencryptoki. It occurs when the `CKA_ID` attribute type is being added to a new attribute list that's being populated. See [this line](https://github.com/opencryptoki/opencryptoki/blob/b74736d144e9215aed47e75bf2fdd54aef0c501c/usr/sbin/p11sak/p11sak.c#L3000). When the type is `CKA_ID` (0x101), `value_len` will be 0, which leads to a `malloc(0)` call.

There was something similar in the provider code as well, though unfortunately I do not recall its exact location now.

Reply:
Another point is that the `malloc` with zero length cannot be "fixed" - the ID field is optional, and when it's being processed the length _will_ be zero. I've fixed the failure check to be length-aware - if the input length was zero, it will not error out anymore. A `memcpy` with zero size (and a `NULL` source buffer) is effectively a no-op, so that's okay too. All the p11sak testcases pass correctly with this fix.

Reply:
I think https://github.com/opencryptoki/opencryptoki/blob/b74736d144e9215aed47e75bf2fdd54aef0c501c/usr/sbin/p11sak/p11sak.c#L3000 should be fixed to not attempt to call malloc at all when the length is zero. Just an if around the malloc, the NULL check and the memcpy, thats it.

I generally don't like code that depends on malloc returning non-NULL on a zero length. This should be handled and malloc should not be called with a zero length.